src/Components/Components/src/Routing/SupplyParameterFromQueryValueProvider.cs

```diff
@@ -0,0 +1,134 @@
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Routing;
+
+internal sealed class SupplyParameterFromQueryValueProvider(NavigationManager navigationManager) : ICascadingValueSupplier, IDisposable
+{
+    private QueryParameterValueSupplier? _queryParameterValueSupplier;
+    private HashSet<ComponentState>? _subscribers;
+    private HashSet<ComponentState>? _pendingSubscribers;
+    private bool _isSubscribedToLocationChanges;
+    private string? _lastUri;
+
+    public bool IsFixed => false;
+
+    public bool CanSupplyValue(in CascadingParameterInfo parameterInfo)
+        => parameterInfo.Attribute is SupplyParameterFromQueryAttribute;
+
+    public object? GetCurrentValue(in CascadingParameterInfo parameterInfo)
+    {
+        TryUpdateQueryParameters();
+
+        var attribute = (SupplyParameterFromQueryAttribute)parameterInfo.Attribute; // Must be a valid cast because we check in CanSupplyValue
+        var queryParameterName = attribute.Name ?? parameterInfo.PropertyName;
+        return _queryParameterValueSupplier.GetQueryParameterValue(parameterInfo.PropertyType, queryParameterName);
+    }
+
+    public void Subscribe(ComponentState subscriber, in CascadingParameterInfo parameterInfo)
+    {
+        if (_pendingSubscribers?.Count > 0 || (TryUpdateQueryParameters() && _isSubscribedToLocationChanges))
+        {
+            // This branch is only taken if there's a pending OnLocationChanged event for the current Uri that we're already subscribed to.
+            // We'll add the _pendingSubscribers to _subscribers and clear _pendingSubscribers during the upcoming call to OnLocationChanged.
+            _pendingSubscribers ??= new();
+            _pendingSubscribers.Add(subscriber);
+            return;
+        }
+
+        _subscribers ??= new();
+        _subscribers.Add(subscriber);
+        SubscribeToLocationChanges();
+    }
+
+    public void Unsubscribe(ComponentState subscriber, in CascadingParameterInfo parameterInfo)
+    {
+        // ICascadingValueSupplier is internal, and Subscribe should always precede Unsubscribe.
+        Debug.Assert(_subscribers is not null);
+        _subscribers.Remove(subscriber);
+        _pendingSubscribers?.Remove(subscriber);
+
+        if (_subscribers.Count == 0 && _pendingSubscribers is null or { Count: 0 })
+        {
+            UnsubscribeFromLocationChanges();
+        }
+    }
+
+    [MemberNotNull(nameof(_queryParameterValueSupplier))]
+    private bool TryUpdateQueryParameters()
+    {
+        _queryParameterValueSupplier ??= new();
+
+        // Router.OnLocationChanged calls GetCurrentValue before our own OnLocationChanged handler,
+        // so we have to compare strings rather than rely on a bool set in OnLocationChanged.
+        if (navigationManager.Uri == _lastUri)
+        {
+            return false;
+        }
```

Comment:
Do we want to be more "precise" here? As in, do we want to find the querystring and compare that bit? We don't care if the path or hash change, do we?

Reply:
Seems like we can call `GetQueryString` on both instances and compare that instead of the full URL.

Reply:
https://github.com/dotnet/aspnetcore/pull/53443/files#diff-71e3430f05198494d6d314c3ffcd4ee38d0ba908458544901a96a0b66824dbe4R78

Reply:
If we're doing this, we might even prefer to do `Object.ReferenceEquals(old, new) || GetQueryString(old) == GetQueryString(new)` since in most cases if it hasn't changed, it will be the same string instance. That saves us from the work of finding the `?` etc.

Reply:
That's fine by me. We call `GetQueryString` later on in the code, so we might as well call it earlier and keep the value in a local.

Reply:
The scenario of a component binding a parameter to the query string and interpreting it the same way regardless of the endpoint is a little strange. And being less precise just means overreporting possible parameter changes which still happens if only an irrelevant part of the query string changes which I suspect is even more common.

`if (navigationManager.Uri == _lastUri)` already short-circuits on reference equality, and we still need to compare the full URI so we know whether the components attempting to subscribe have seen the current Uri already regardless of whether or not the query component of the URI changed. Even if the query string has changed, those components do not need to be notified about a URI they have already observed.

The fix would have to look something like [this](https://github.com/dotnet/aspnetcore/commit/52c7e37462b1c35adfe266f7917fc0e190583911), and I don't think it's worth that level of extra complexity. If someone has a far simpler way to do this, I'm all ears.

Reply:
@halter73 you've thought more deeply about this than the rest of us recently so would be best placed to make a call on what you think is the best tradeoff (@javiercn if you feel strongly we should do a specific thing, please say so!)

Reply:
No, I'll trust @halter73 makes the right calls